build-system/tasks/e2e/selenium-webdriver-controller.js

```diff
@@ -82,6 +90,29 @@ class SeleniumWebDriverController {
     };
   }
 
+  async waitForElementProperty(handle, property, value) {
+    value = value.toString();
+    const element = handle.getElement();
+    await this.driver.wait(async() => {
+      let actualValue = await element.getAttribute(property);
+      // certain attributes will return true or null, and
+      // others will return a string
+      //https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/index_exports_WebElement.html#getAttribute
+      actualValue = actualValue != null ? actualValue.toString() : null;
+      return actualValue === value;
+    }, 10000, `Error waiting for element property '${property}'` +
+      ` to be ${value}`);
+  }
+
+  async waitForElementRect(handle, rect) {
```

Comment:
Is this necessary? When I change one of my tests to have an invalid rect value, I get a message like:

```
      Uncaught AssertionError: expected { Object (x, y, ...) } to have property 'width' of 600, but got 800
      + expected - actual

      -800
      +600
```

We should solve the assertion message more generally (i.e. for all calls to `include`) rather than just for just for rects or just for properties. I already added code for this by capturing the last error message when waiting for a condition, and then throwing that when we hit the timeout. The approach I used for that is not super robust, so if that isn't working, we should fix / improve that rather than trying to add error handling here.
 
It looks like this also hard codes the a timeout, which does not match the test timeout. I do not think that is something we should be doing.

Reply:
Good point about hard coding the timeout.

I don't get an AssertionError when I run tests locally. Instead I get timeout errors that look like this.
```
  20) AMP list load-more=auto
        
         should load more items on scroll:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/usr/local/google/home/esth/src/ampproject/amphtml/extensions/amp-list/0.1/test-e2e/test-load-more-auto.js)
  

Executed 41 of 48 (Skipped 7) 20 FAILED 
```
Where are you throwing the last error message before a timeout message is thrown?

Reply:
The code is here:

https://github.com/ampproject/amphtml/blob/086bbe497f3fce219d35ddcc8a636d0061254297/build-system/tasks/e2e/describes-e2e.js#L109-L115

I set a timeout for just before the test itself times out (by subtracting 1ms), but I don't think that is very robust. Maybe we can move the error reporting logic into `afterEach`, if that runs when you timeout. If not, perhaps we can implement our own custom test timeout logic rather than using the test runner's.